docs/rfc/rfc-017-abci++-vote-extension-propag.md

```diff
@@ -0,0 +1,568 @@
+          itself as caught-up. If a validator is indeed late for height *h*, it is useless (although
+          correct) for it to call `PrepareProposal`, or `ExtendVote`, since the block is already decided.
+        - Moreover, some use cases require pretty sizeable vote extensions, which would result in an
+          important waste of space in the blockchain.
+
+- **Solution 2.** *Skip* propose *step in Tendermint algorithm*.
+
+    This solution consists in modifying the Tendermint algorithm to skip the *send proposal* step in
+    heights where the node does not have the required vote extensions to populate the call to
+    `PrepareProposal`. The main idea behind this is that it should only happen when the validator is late
+    and, therefore, up-to-date validators have already proposed (and decided) for that height.
+    A small variation of this solution is, rather than skipping the *send proposal* step, the validator
+    sends a special *empty* or *bottom* (⊥) proposal to signal other nodes that it is not ready to propose
+    at (any round of) the current height.
+
+    The appeal of this solution is its simplicity. A possible implementation does not need to extend
+    the data structures, or change the current catch-up mechanisms implemented in the blocksync or
+    in the consensus reactor. When we lack the needed information (vote extensions), we simply rely
+    on another correct validator to propose a valid block in other rounds of the current height.
+
+    However, this solution can be attacked by a byzantine node in the network in the following way.
+    Let us consider the following scenario:
+
+    - all validators in *valset<sub>h</sub>* send out precommit messages, with vote extensions,
+      for height *h*, round 1, roughly at the same time,
+    - all those precommit messages contain non-`nil` precommit votes, which vote for block *b*
+    - all validators in *valset<sub>h</sub>* then wait until they gather enough precommit
+      messages for height *h*, round 1, in order to decide in height *h*,
+    - all precommit messages sent in height *h*, round 1, and all messages sent in
+      height *h*, round *r > 1* get delayed indefinitely,
+    - an intermediate (malicious) full node *m* manages to receive block *b*, and gather more than
+      *2n<sub>h</sub>/3* precommit messages for height *h*, round 1,
+    - one way or another, the solution should have a mechanism for a full node to tell another full
+      node it is late; this mechanism should somehow include a message type containing － at the very
+      least － the decided block and a commit (not necessarily an extended commit) for *h*;
+      let us call this message type *catch-up message*,
+    - node *m* uses the gathered precommit messages to build a commit for height *h*, round 1,
+    - in order to convince them they are late, node *m* sends all full nodes a catch-up message with
+      block *b*, and the commit for height *h*, round 1,
```

Comment:
Let's say there is a period of network instability. We usually model this by assuming that the malicious node has control over the  network's performance (it's actually a worst-case scenario).

If some full node nevertheless receives enough precommits to decide by itself, but also receives the catch-up message from _m_, it'll be very hard (impossible?) to prove _m_ is misbehaving, since we can't distinguish that from a correct node legitimately believing other nodes are late (due to e.g. receiving some delayed message).
Bear in mind that the information conveyed in the catch-up message, namely, the commit and the block is correct. The only thing that _m_ forges is its perception that other nodes are late or not.

On the other hand, as the text says, if _m_ is ever able to reach all full nodes with its catch-up message, before enough precommit votes, then all full nodes (validators included) fall into the problem of advancing to _h+1_ and expecting someone else to propose.